File: My-Leetcode/54. 螺旋矩阵【数组】.cpp
#include "headfile.h"
using namespace std;
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        if(matrix.size()==0||matrix[0].size()==0) return res;
        vector<vector<int> > judge(matrix.size(),vector<int>(matrix[0].size(),0)); // 定义一个判断的矩阵。

        int x= 0,y=0;// 定义起始位置
        res.push_back(matrix[x][y]);
        judge[0][0] =1;
        int flag = 1;
        while(flag)
        {
            flag = 0; //如果都没得走，结束循环
            // 先向右走
            while(is_legal(x,y+1,judge))
                y++,res.push_back(matrix[x][y]),flag=1,judge[x][y]=1;

            // 向下走
            while(is_legal(x+1,y,judge))
                x++,res.push_back(matrix[x][y]),flag=1,judge[x][y]=1;

            //向左走
            while(is_legal(x,y-1,judge))
                y--,res.push_back(matrix[x][y]),flag=1,judge[x][y]=1;

            while(is_legal(x-1,y,judge))
                x--,res.push_back(matrix[x][y]),flag=1,judge[x][y]=1;
        }
        return res;
    }
    bool is_legal(int x,int y, vector<vector<int> >& judge)
    {
        int limit_x = judge.size();
        int limit_y = judge[0].size();
        if(x>=0 && x<limit_x && y>=0 && y<limit_y &&judge[x][y]==0)
            return true;
        return false;
    }
};

File: My-Leetcode/Python/54. 螺旋矩阵【数组】.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        res = []
        if len(matrix) <=0 or len(matrix[0]) <= 0:
            return res
        self.N, self.M = len(matrix), len(matrix[0])
        dist_seq = [[0, 1], [1, 0], [0, -1], [-1, 0]]

        i, j, dist = 0, 0, 0
        while True:
            res.append(matrix[i][j])
            matrix[i][j] = "/"
            next_i, next_j = i + dist_seq[dist][0], j + dist_seq[dist][1]
            if not self.legal(next_i,next_j,matrix):
                if len(res) == self.N * self.M:
                    return res
                # 换方向
                dist = (dist + 1) % len(dist_seq)
                next_i, next_j = i + dist_seq[dist][0], j + dist_seq[dist][1]
            i, j = next_i, next_j
    
    def legal(self, i , j, matrix):
        if i<0 or i>=self.N or j<0 or j>=self.M or matrix[i][j] == "/":
            return False
        return True